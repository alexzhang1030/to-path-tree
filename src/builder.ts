import type { NodeItem, ParseResults, TreeNode } from './types'
import { DEFAULT_SEP, ROOT_NAME } from './constants'
import { genRoot, parsePath } from './utils'

export class PathTreeBuilder<T extends Record<string, unknown>> {
  #root: TreeNode<T> = genRoot()
  #mapping = new Map<string, ParseResults<T>>()

  #sep
  #getData
  #autoResolveDir

  /**
   * @param options
   * @param options.sep - default `/`
   * @param options.getData - function to get data from node
   * @param options.autoResolveDir - will auto resolve directory when adding file path, but not added dir path, default `true`
   */
  constructor(options?: {
    sep?: string
    getData?: (node: NodeItem<T>) => T
    autoResolveDir?: boolean
  }) {
    const { sep = DEFAULT_SEP, getData, autoResolveDir = true } = options ?? {}

    this.#sep = sep
    this.#getData = getData
    this.#autoResolveDir = autoResolveDir
  }

  /**
   *
   * @param path
   * @param userData pass this will override default getData passed in constructor
   */
  addPath(path: string, userData?: T) {
    const nodeItem = parsePath(
      path,
      this.#sep,
      this.#root,
      userData
        ? () => userData
        : this.#getData,
    )
    this.#mapping.set(nodeItem.node.path, nodeItem)
    if (this.#autoResolveDir)
      this.#resolveDir(nodeItem)
  }

  #resolveDir(node: ParseResults<T>) {
    let dirNode: ParseResults<T>['node'] | undefined = node.node
    while (dirNode.parent) {
      dirNode = dirNode?.parent
      if (!this.#mapping.has(dirNode.path)) {
        this.#mapping.set(dirNode.path, {
          node: dirNode,
          type: 'directory',
        })
      }
    }
  }

  removePath(path: string) {
    const nodeItem = this.#mapping.get(path)
    if (!nodeItem)
      return
    const { node, type } = nodeItem
    if (type === 'file')
      node.parent?.items.splice(node.parent.items.indexOf(node), 1)
    else
      Reflect.deleteProperty(node.parent!.subDirectory!, node.name)
  }

  /**
   *
   * @param path path to get node
   * @param _isFile pass isFile for better type inference
   */
  getNode<const IF extends boolean>(path: string, _isFile: IF = true as IF) {
    const node = this.#mapping.get(path)
    if (!node)
      return null
    return node.node as IF extends true ? NodeItem<T> : TreeNode<T>
  }

  getItems(path: string) {
    if (path === ROOT_NAME)
      return this.#root.items
    const node = this.#mapping.get(getStrictPath(path))
    if (!node)
      return []
    return node.type === 'file'
      ? []
      : node.node.items
  }

  getSubDirectories(path: string) {
    if (path === ROOT_NAME)
      return this.#root.subDirectory
    const node = this.getNode(getStrictPath(path), false)
    if (!node)
      return null
    return node.subDirectory
  }

  get tree() {
    return this.#root
  }
}

function getStrictPath(path: string) {
  return path.startsWith('/') ? path : `/${path}`
}
