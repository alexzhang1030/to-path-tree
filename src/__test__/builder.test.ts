import { PathTreeBuilder } from '@/builder'
import { ROOT_NAME } from '..'

it('simple builder', () => {
  const builder = new PathTreeBuilder()
  builder.addPath('/a/')
  builder.addPath('/a/b/')
  builder.addPath('/a/b/index.ts')
  builder.addPath('/a/b/foo.ts')
  expect(builder.tree).toMatchSnapshot('simple-builder add path')
  builder.removePath('/a/b/')
  builder.removePath('/a/b/foo.ts')
  builder.removePath('/a/b/baz.ts') // remove non-exist path
  expect(builder.tree).toMatchSnapshot('simple-builder remove path')
})

it('get items w/autoResolveDir = false', () => {
  const builder = new PathTreeBuilder({ autoResolveDir: false })
  builder.addPath('/a/')
  builder.addPath('/a/index.ts')
  expect(builder.tree).toMatchSnapshot()
  expect(builder.getItems('/a/')).toMatchSnapshot()
  // get items from file, should return []
  expect(builder.getItems('/a/index.ts')).toEqual([])
  // get non-exist path, should return []
  expect(builder.getItems('/a/b/')).toEqual([])
})

it('get items w/autoResolveDir = true', () => {
  const builder = new PathTreeBuilder()
  builder.addPath('src/index.ts')
  builder.addPath('src/a/index.ts')
  builder.removePath('src/index.ts')
  builder.addPath('src/b/')
  expect(builder.getItems('src/a/')).toMatchSnapshot()
  expect(builder.getSubDirectories('src/')).toMatchSnapshot()
  // get root items
  expect(builder.getItems(ROOT_NAME)).toMatchSnapshot()
  // get root sub directories
  expect(builder.getSubDirectories(ROOT_NAME)).toMatchSnapshot()
  // get sub directories but pass file path
  expect(builder.getSubDirectories('src/index.ts')).toBeNull()
})

it('custom separator', () => {
  const builder = new PathTreeBuilder({ sep: '$' })
  builder.addPath('$a$')
  builder.addPath('$a$b$')
  builder.addPath('$a$b$index.ts')
  builder.addPath('$a$b$foo.ts')
  builder.addPath('$a$z$index.ts')
  expect(builder.tree).toMatchSnapshot('custom-separator add path')
  builder.removePath('$a$b$')
  builder.removePath('$a$b$foo.ts')
  builder.removePath('$a$b$baz.ts') // remove non-exist path
  expect(builder.tree).toMatchSnapshot('custom-separator remove path')
})

it('get data', () => {
  const builder = new PathTreeBuilder({
    getData: node => ({ data: node.ext }),
  })
  builder.addPath('/a/index.ts')
  // override default
  builder.addPath('/a/foo.ts', { data: 'hello world' })
  const node = builder.getNode('/a/index.ts', true)
  expect(node?.data).toStrictEqual({
    data: 'ts',
  })
  expect(builder.getNode('/a/foo.ts', true)?.data).toStrictEqual({
    data: 'hello world',
  })
  // get non-exist path, should return null
  expect(builder.getNode('/a/b/index.ts')).toBeNull()
})
