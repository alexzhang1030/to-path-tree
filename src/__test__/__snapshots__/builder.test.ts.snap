// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom separator > custom-separator add path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a$",
      "relativePath": "a$",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "$a$b$index.ts",
            },
            {
              "data": undefined,
              "ext": "ts",
              "file": "foo.ts",
              "filename": "foo",
              "isEntry": false,
              "parent": [Circular],
              "path": "$a$b$foo.ts",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/a$b$",
          "relativePath": "b$",
          "relativePathName": "b",
          "subDirectory": {},
        },
        "z": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "$a$z$index.ts",
            },
          ],
          "name": "z",
          "parent": [Circular],
          "path": "/a$z$",
          "relativePath": "z$",
          "relativePathName": "z",
          "subDirectory": {},
        },
      },
    },
  },
}
`;

exports[`custom separator > custom-separator remove path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a$",
      "relativePath": "a$",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "$a$b$index.ts",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/a$b$",
          "relativePath": "b$",
          "relativePathName": "b",
          "subDirectory": {},
        },
        "z": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "$a$z$index.ts",
            },
          ],
          "name": "z",
          "parent": [Circular],
          "path": "/a$z$",
          "relativePath": "z$",
          "relativePathName": "z",
          "subDirectory": {},
        },
      },
    },
  },
}
`;

exports[`get items w/autoResolveDir = false 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [
        {
          "data": undefined,
          "ext": "ts",
          "file": "index.ts",
          "filename": "index",
          "isEntry": true,
          "parent": [Circular],
          "path": "/a/index.ts",
        },
      ],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {},
    },
  },
}
`;

exports[`get items w/autoResolveDir = false 2`] = `
[
  {
    "data": undefined,
    "ext": "ts",
    "file": "index.ts",
    "filename": "index",
    "isEntry": true,
    "parent": {
      "items": [Circular],
      "name": "a",
      "parent": {
        "items": [],
        "name": "root",
        "path": "/",
        "relativePath": "/",
        "relativePathName": "",
        "subDirectory": {
          "a": [Circular],
        },
      },
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {},
    },
    "path": "/a/index.ts",
  },
]
`;

exports[`get items w/autoResolveDir = true 1`] = `
[
  {
    "data": undefined,
    "ext": "ts",
    "file": "index.ts",
    "filename": "index",
    "isEntry": true,
    "parent": {
      "items": [Circular],
      "name": "a",
      "parent": {
        "items": [],
        "name": "src",
        "parent": {
          "items": [],
          "name": "root",
          "path": "/",
          "relativePath": "/",
          "relativePathName": "",
          "subDirectory": {
            "src": [Circular],
          },
        },
        "path": "/src/",
        "relativePath": "src/",
        "relativePathName": "src",
        "subDirectory": {
          "a": [Circular],
          "b": {
            "items": [],
            "name": "b",
            "parent": [Circular],
            "path": "/src/b/",
            "relativePath": "b/",
            "relativePathName": "b",
            "subDirectory": {},
          },
        },
      },
      "path": "/src/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {},
    },
    "path": "src/a/index.ts",
  },
]
`;

exports[`get items w/autoResolveDir = true 2`] = `
{
  "a": {
    "items": [
      {
        "data": undefined,
        "ext": "ts",
        "file": "index.ts",
        "filename": "index",
        "isEntry": true,
        "parent": [Circular],
        "path": "src/a/index.ts",
      },
    ],
    "name": "a",
    "parent": {
      "items": [],
      "name": "src",
      "parent": {
        "items": [],
        "name": "root",
        "path": "/",
        "relativePath": "/",
        "relativePathName": "",
        "subDirectory": {
          "src": [Circular],
        },
      },
      "path": "/src/",
      "relativePath": "src/",
      "relativePathName": "src",
      "subDirectory": [Circular],
    },
    "path": "/src/a/",
    "relativePath": "a/",
    "relativePathName": "a",
    "subDirectory": {},
  },
  "b": {
    "items": [],
    "name": "b",
    "parent": {
      "items": [],
      "name": "src",
      "parent": {
        "items": [],
        "name": "root",
        "path": "/",
        "relativePath": "/",
        "relativePathName": "",
        "subDirectory": {
          "src": [Circular],
        },
      },
      "path": "/src/",
      "relativePath": "src/",
      "relativePathName": "src",
      "subDirectory": [Circular],
    },
    "path": "/src/b/",
    "relativePath": "b/",
    "relativePathName": "b",
    "subDirectory": {},
  },
}
`;

exports[`get items w/autoResolveDir = true 3`] = `[]`;

exports[`get items w/autoResolveDir = true 4`] = `
{
  "src": {
    "items": [],
    "name": "src",
    "parent": {
      "items": [],
      "name": "root",
      "path": "/",
      "relativePath": "/",
      "relativePathName": "",
      "subDirectory": [Circular],
    },
    "path": "/src/",
    "relativePath": "src/",
    "relativePathName": "src",
    "subDirectory": {
      "a": {
        "items": [
          {
            "data": undefined,
            "ext": "ts",
            "file": "index.ts",
            "filename": "index",
            "isEntry": true,
            "parent": [Circular],
            "path": "src/a/index.ts",
          },
        ],
        "name": "a",
        "parent": [Circular],
        "path": "/src/a/",
        "relativePath": "a/",
        "relativePathName": "a",
        "subDirectory": {},
      },
      "b": {
        "items": [],
        "name": "b",
        "parent": [Circular],
        "path": "/src/b/",
        "relativePath": "b/",
        "relativePathName": "b",
        "subDirectory": {},
      },
    },
  },
}
`;

exports[`simple builder > simple-builder add path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "/a/b/index.ts",
            },
            {
              "data": undefined,
              "ext": "ts",
              "file": "foo.ts",
              "filename": "foo",
              "isEntry": false,
              "parent": [Circular],
              "path": "/a/b/foo.ts",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/a/b/",
          "relativePath": "b/",
          "relativePathName": "b",
          "subDirectory": {},
        },
      },
    },
  },
}
`;

exports[`simple builder > simple-builder remove path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {},
    },
  },
}
`;
